frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -216,71 +210,39 @@ private void refreshModel(
-        }
-
-        var repositorySnapshot = discoverRunResult.get();
-
-        if (repositorySnapshot.getRootBranches().size() == 0) {
-          GuiUtils.invokeLaterIfNeeded(
-              () -> setTextForEmptyTable(
-                  format(getString("string.GitMachete.EnhancedGraphTable.empty-table-text.cant-discover-layout"),
-                      macheteFilePath.toString())),
-              NON_MODAL);
-          return;
-        }
-
-        var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
-        var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+  @Override
+  public void queueDiscover(Path macheteFilePath, @UI Runnable doOnUIThreadWhenReady) {
+    new GitMacheteRepositoryDiscoverer(
+        project,
+        getGitRepositorySelectionProvider(),
+        getUnsuccessfulDiscoverMacheteFilePathConsumer(),
+        getSuccessfulDiscoverRepositoryConsumer(doOnUIThreadWhenReady))
+            .enqueue(macheteFilePath);
+  }
 
-        if (branchLayout == null) {
-          LOG.error("Can't get branch layout from repository snapshot");
-          return;
-        }
+  @UIEffect
+  private Consumer<IGitMacheteRepositorySnapshot> getSuccessfulDiscoverRepositoryConsumer(@UI Runnable doOnUIThreadWhenReady) {
+    return (IGitMacheteRepositorySnapshot repositorySnapshot) -> GuiUtils.invokeLaterIfNeeded(
+        () -> {
+          gitMacheteRepositorySnapshot = repositorySnapshot;
+          queueRepositoryUpdateAndModelRefresh(doOnUIThreadWhenReady);
+
+          VcsNotifier.getInstance(project)
+              .notifyInfo(getString("string.GitMachete.EnhancedGraphTable.automatic-discover.success-message"));
+        },
+        NON_MODAL);
+  }
 
-        try {
-          branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true);
-
-          GuiUtils.invokeLaterIfNeeded(
-              () -> {
-                gitMacheteRepositorySnapshot = repositorySnapshot;
-                queueRepositoryUpdateAndModelRefresh(doOnUIThreadWhenReady);
-
-                VcsNotifier.getInstance(project)
-                    .notifyInfo(getString("string.GitMachete.EnhancedGraphTable.automatic-discover.success-message"));
-              },
-              NON_MODAL);
-        } catch (BranchLayoutException exception) {
-          GuiUtils.invokeLaterIfNeeded(() -> VcsNotifier.getInstance(project).notifyError(
-              getString("string.GitMachete.EnhancedGraphTable.automatic-discover.cant-discover-layout-error-title"),
-              exception.getMessage() != null ? exception.getMessage() : ""), NON_MODAL);
-        }
-      }
-    }.queue();
+  @UIEffect
+  private Consumer<Path> getUnsuccessfulDiscoverMacheteFilePathConsumer() {
+    return (Path macheteFilePath) -> GuiUtils.invokeLaterIfNeeded(
+        () -> setTextForEmptyTable(
+            format(getString("string.GitMachete.EnhancedGraphTable.empty-table-text.cant-discover-layout"),
```

Comment:
Nit: I can see there's one message string that uses `cannot-open` rather than `cant`... pls make consistent